src/sequence_batch_scheduler.cc

```diff
@@ -110,12 +108,143 @@ SequenceBatchScheduler::Create(
   // runner. This is at least 1 even if the model doesn't support
   // batching.
   const size_t model_batch_size = std::max(1, config.max_batch_size());
-  size_t seq_slot_cnt = model_batch_size;
+  sched->seq_slot_cnt_ = model_batch_size;
   if (config.sequence_batching().has_oldest()) {
-    seq_slot_cnt =
+    sched->seq_slot_cnt_ =
         config.sequence_batching().oldest().max_candidate_sequences();
   }
 
+  // Create batchers.
+  RETURN_IF_ERROR(sched->CreateBatchers(model->BackgroundInstances()));
+
+  // Create a reaper thread that watches for idle sequences. Run the
+  // reaper a lower priority.
+  SequenceBatchScheduler* raw = sched.release();
+
+  raw->reaper_thread_exit_ = false;
+  raw->reaper_thread_.reset(
+      new std::thread([raw]() { raw->ReaperThread(10 /* nice */); }));
+
+  scheduler->reset(raw);
+
+  return Status::Success;
+}
+
+Status
+SequenceBatchScheduler::Update()
+{
+  // Find added and removed instances for this update.
+  std::vector<std::shared_ptr<TritonModelInstance>> added_instances;
+  std::unordered_set<TritonModelInstance*> removed_instances;
+  InstancesDiff(&added_instances, &removed_instances);
+
+  // Batchers should be destroyed outside the lock.
+  std::vector<std::unique_ptr<SequenceBatch>> removed_batchers;
+
+  {
+    std::unique_lock<std::mutex> lk(mu_);
+
+    // Add the new batchers.
+    RETURN_IF_ERROR(CreateBatchers(added_instances));
+
+    // All sequence slots of the removed instances are pending to be removed.
+    pending_removal_seq_slots_.clear();
+    for (auto& instance : removed_instances) {
+      pending_removal_seq_slots_.emplace(instance, seq_slot_cnt_);
```

Comment:
Should use the count captured in the corresponding SequenceBatch object. It is probably not the case here, but note that `seq_slot_cnt_` is not proper indicator of the slots associated to the given instance, which can be dangerous if `seq_slot_cnt_` may get updated after the SequenceBatch is constructed.

Reply:
This is a good point. Updated: [Use slot count when batcher is created during removal](https://github.com/triton-inference-server/core/pull/207/commits/bde56a383342f6cb57d00e03832f8bfd96575f2e)